### PR TITLE
WIP: Undo/Redo Command implementation

### PR DESCRIPTION
**Pull request by saad-haider-pp**
This PR brings the implementation for the undo and redo command.
The tests and documentation haven't been done yet, but I felt it important to merge this quickly to prevent future conflicts

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/Command.java [hunk 1]
```diff
@@ -16,21 +16,25 @@
...
+     * @param model {@code Model} which contains the model which the command operates on
      * @throws CommandException If an error occurs during command execution.
      */
-    public abstract CommandResult execute(Model model) throws CommandException;
-
-    protected void requireDeliveryModel(Model model) throws CommandException {
-        if (!(model instanceof DeliveryModel)) {
-            throw new CommandException(MESSAGE_INCORRECT_MODEL);
+    public CommandResult execute(Model model) throws CommandException {
+        if (model instanceof InventoryModel) {
```
Comment (Wincenttjoi) [inline src/main/java/seedu/address/logic/commands/Command.java]: Can just check why are the types opposite?

Comment (saad-haider-pp) [reply src/main/java/seedu/address/logic/commands/Command.java]: What do you mean

File: src/main/java/seedu/address/logic/commands/Command.java [hunk 2]
```diff
@@ -16,21 +16,25 @@
...
+            return execute((InventoryModel) model, new DeliveryModelManager());
+        } else {
+            return execute(new InventoryModelManager(), (DeliveryModelManager) model);
```
Comment (Wincenttjoi) [inline src/main/java/seedu/address/logic/commands/Command.java:26]: Sorry I meant these few lines above, why casting and the actual model manager is opposite?

File: src/main/java/seedu/address/logic/commands/Command.java (hunk 1 above)
Comment (Wincenttjoi) [reply src/main/java/seedu/address/logic/commands/Command.java]: Ah nvm my mistake, understood whats happening alrd

**Comment by Wincenttjoi** [comment]
We PR once u done the test cases first?

**Comment by saad-haider-pp** [comment]
@Wincenttjoi I'm having some issues with the test cases (they work individually but don't work when the whole test suite is run. I've been googling but still am not sure where the issue is), don't think they'll be done any time soon. 

Maybe PR first so that the rest of development can work on top of it (for instance the new delivery commands will need to take this into consideration now)

Edit: don't merge this yet

File: src/main/java/seedu/address/logic/commands/Command.java (hunk 2 above)
Comment (Choo-Xing-Yu) [inline src/main/java/seedu/address/logic/commands/Command.java]: @halcon-blanco as discussed in telegram, we should probably create a wrapper class so we can make use of polymorphism instead of using instanceof

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/logic/commands/Command.java]: But for now let's implement the rest on top of this first

File: src/main/java/seedu/address/logic/commands/itemcommand/ItemListCommand.java [hunk 3]
```diff
@@ -19,10 +18,9 @@
 
 
     @Override
-    public CommandResult execute(Model model) throws CommandException {
-        requireNonNull(model);
-        requireInventoryModel(model);
-        InventoryModel inventoryModel = (InventoryModel) model;
+    public CommandResult execute(InventoryModel inventoryModel, DeliveryModel deliveryModel) {
+        requireNonNull(inventoryModel);
```
Comment (Choo-Xing-Yu) [inline src/main/java/seedu/address/logic/commands/itemcommand/ItemListCommand.java]: Should it check for deliveryModel to be nonNull too?

File: src/main/java/seedu/address/logic/commands/itemcommand/ItemFindCommand.java [hunk 4]
```diff
@@ -30,10 +29,8 @@ public ItemFindCommand(ItemContainsKeywordsPredicate predicate) {
     }
 
     @Override
-    public CommandResult execute(Model model) throws CommandException {
-        requireNonNull(model);
-        requireInventoryModel(model);
-        InventoryModel inventoryModel = (InventoryModel) model;
+    public CommandResult execute(InventoryModel inventoryModel, DeliveryModel deliveryModel) {
+        requireNonNull(inventoryModel);
```
Comment (Choo-Xing-Yu) [inline src/main/java/seedu/address/logic/commands/itemcommand/ItemFindCommand.java]: Should it check for deliveryModel to be nonNull too?

File: src/main/java/seedu/address/logic/commands/Command.java (hunk 2 above)
Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/logic/commands/Command.java]: @halcon-blanco i have a suggestion. We should make use of DeliveryCommand and ItemCommand abstract class to have a abstract method execute that takes in its respective model. This will adhere to OpenClosed principle better

File: src/main/java/seedu/address/logic/commands/itemcommand/ItemListCommand.java (hunk 3 above)
Comment (saad-haider-pp) [reply src/main/java/seedu/address/logic/commands/itemcommand/ItemListCommand.java]: I know we're changing this, but to answer this, I didn't think it necessary to check deliveryModel since it wasn't being used

File: src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java [hunk 5]
```diff
@@ -108,6 +108,8 @@ public void removeDelivery(Delivery key) {
      * Copies the current {@code InventoryBook} in the state list.
      */
     public void commit() {
+        assert deliveryBookStatePointer < deliveryBookStateList.size();
+
```
Comment (xnoobftw) [inline src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java:112]: Should this throw UndoRedoLimitReachedException instead of being an assertion?

File: src/main/java/seedu/address/model/inventorymodel/InventoryBook.java [hunk 6]
```diff
@@ -112,6 +112,8 @@ public void removeItem(Item key) {
      * Copies the current {@code InventoryBook} in the state list.
      */
     public void commit() {
+        assert inventoryBookStatePointer < inventoryBookStateList.size();
+
```
Comment (xnoobftw) [inline src/main/java/seedu/address/model/inventorymodel/InventoryBook.java:116]: Same here!

File: src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java [hunk 7]
```diff
@@ -108,6 +108,8 @@ public void removeDelivery(Delivery key) {
...
         if (!deliveryBookStateList.isEmpty()) {
             deliveryBookStateList = deliveryBookStateList.subList(0, deliveryBookStatePointer + 1);
         }
```
Comment (Choo-Xing-Yu) [inline src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java]: I feel that similarly to the command Hisotry there should be a limit of how many times u can redo incase of memory overflow

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java]: Like limit the number of state stateList holds

Comment (saad-haider-pp) [reply src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java]: Good point. Will add it

File: src/main/java/seedu/address/model/inventorymodel/InventoryBook.java [hunk 8]
```diff
@@ -117,6 +118,11 @@ public void commit() {
         if (!inventoryBookStateList.isEmpty()) {
             inventoryBookStateList = inventoryBookStateList.subList(0, inventoryBookStatePointer + 1);
         }
+        if (inventoryBookStateIsFull()) {
+            inventoryBookStateList.remove(0);
+            inventoryBookStatePointer--;
+        }
+
```
Comment (Choo-Xing-Yu) [inline src/main/java/seedu/address/model/inventorymodel/InventoryBook.java:128]: In my experience with CommandHistory's previousCommand(), be wary of the corner case if suppose:
you have 2 states, state#1, state#2

undo() should return you to state#2 and not state#1.
Could you give that a shot? Maybe either via junit tests or manual testing

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/inventorymodel/InventoryBook.java:128]: This is a complication from having the pointer pointing at the last element of the list

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/inventorymodel/InventoryBook.java:128]: Nvm ignore me i've read through your implementation and realised its different

File: src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java [hunk 9]
```diff
@@ -1,14 +1,21 @@
...
+import java.util.ArrayList;
 import java.util.List;
 
 import javafx.collections.ObservableList;
+import seedu.address.logic.commands.exceptions.UndoRedoLimitReachedException;
 import seedu.address.model.delivery.Delivery;
 import seedu.address.model.delivery.UniqueDeliveryList;
 
 public class DeliveryBook implements ReadOnlyDeliveryBook {
+    private static List<DeliveryBook> deliveryBookStateList = new ArrayList<>();
```
Comment (Choo-Xing-Yu) [inline src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java]: Perhaps instantiate arraylist with a initial capacity to optimize performance (remember from 2040s they taught us that arraylist copies over and doubles in size everytime its limit is hit)

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java]: Eg  private static List<DeliveryBook> deliveryBookStateList = new ArrayList<>(MODEL_STATE_LIMIT);

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java]: and i feel that it shouldn't be a static variable simply because we may use different instantiation of deliverybooks in testing as well

File: src/main/java/seedu/address/model/inventorymodel/InventoryBook.java [hunk 10]
```diff
@@ -101,6 +107,64 @@ public void removeItem(Item key) {
...
+    /**
+     * Copies the current {@code InventoryBook} in the state list.
+     */
+    public void commit() {
+        assert inventoryBookStatePointer < inventoryBookStateList.size();
+
+        if (!inventoryBookStateList.isEmpty()) {
+            inventoryBookStateList = inventoryBookStateList.subList(0, inventoryBookStatePointer + 1);
+        }
```
Comment (Choo-Xing-Yu) [inline src/main/java/seedu/address/model/inventorymodel/InventoryBook.java]: i think a better condition would be canRedo() instead

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/inventorymodel/InventoryBook.java]: having it as canRedo optimizes performance as we only really need to do a copy if there exist downstream "commits"

File: src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java (hunk 9 above)
Comment (saad-haider-pp) [reply src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java]: ok let me try

File: src/main/java/seedu/address/model/inventorymodel/InventoryBook.java (hunk 10 above)
Comment (saad-haider-pp) [reply src/main/java/seedu/address/model/inventorymodel/InventoryBook.java]: Makes sense, I shall change it

File: src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java (hunk 9 above)
Comment (saad-haider-pp) [reply src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java]: issue with this is that it becomes hard to make a canRedo() method without adding another pointer, which makes things messy. I think our array limit will never be set beyond a few 1000 so it's not an issue (and if it is ever changed, the code can be refactored then)

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java]: ArrayList.size() != Capacity btw @halcon-blanco

File: src/main/java/seedu/address/model/inventorymodel/InventoryBook.java (hunk 8 above)
Comment (saad-haider-pp) [reply src/main/java/seedu/address/model/inventorymodel/InventoryBook.java:128]: Manual testing shows that it works as expected when the limit is 1 or 2

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/model/inventorymodel/InventoryBook.java:128]: Thanks!!

File: src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java (hunk 9 above)
Comment (saad-haider-pp) [reply src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java]: oh sheez for some reason my github didn't show all your comments

Comment (saad-haider-pp) [reply src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java]: ok so for the static variable, I luckily saw the issue too and just fixed it. as for the capacity != size part, yeah my bad haha still rusty from 2040 times will add it in

Comment (saad-haider-pp) [reply src/main/java/seedu/address/model/deliverymodel/DeliveryBook.java]: ok you can check it out now!

File: src/test/java/seedu/address/logic/commands/UndoCommandTest.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+//        inventoryModel.commit();
+//        inventoryModel.commit();
+//        deliveryModel.commit();
+//        deliveryModel.commit();
+//
+//        assertCommandSuccess(
+//                new UndoCommand(), inventoryModel, MESSAGE_UNDO_ACKNOWLEDGEMENT, inventoryModel);
+//    }
+//
+//}
```
Comment (Choo-Xing-Yu) [inline src/test/java/seedu/address/logic/commands/UndoCommandTest.java:48]: shall merge into master first so others can work with new design prior to implementing undo/redo tests

Comment (saad-haider-pp) [reply src/test/java/seedu/address/logic/commands/UndoCommandTest.java:48]: thanks! shall start work on test cases then

**Review by Choo-Xing-Yu — APPROVED** [review]
LGTM! Good job on implementing redo and undo! Tests will be done in a future PR due to design changes (important to merge early to prevent merge conflict hell)
